Fix some editor crashes

Dentrax opened the pull request:
Fixes #29960
Fixes #29954
Fixes #29963
Fixes #29950
Fixes #29920
Fixes #29767
Fixes #29746
Fixes #29745
Fixes #29673
Fixes #29657
Fixes #29623
Fixes #29622

**IDK:** For #29778 , I waited _~5_ mins and it did not crash.
**IDK:** For #29766 , I waited _~5_ mins and it did not crash.
**IDK:** For #29369 , I waited _~5_ mins and it did not crash.

In addition, Thanks to @qarmin for the well explained test cases.

My second _PR_ on _Godot_. Yay!

Dentrax committed:
fix some crashes

Selene29 commented on scene/main/node.cpp:
```diff
@@ -841,15 +845,19 @@ void Node::set_process_priority(int p_priority) {
 	data.process_priority = p_priority;
 
 	if (is_processing())
+		ERR_FAIL_COND(!data.tree);
```
You need braces here and below, otherwise only the first line after the "if" is part of the if-clause.

Dentrax replied:
Oh, I missed that point. Fixed.

akien-mga replied:
Checking once before the first `if` should be enough.

LinuxUserGD commented:
Seems like @qarmin is the crash king :D

Chaosus commented on scene/resources/visual_shader.cpp:
```diff
@@ -1762,6 +1762,9 @@ void VisualShaderNodeGroupBase::set_inputs(const String &p_inputs) {
 	for (int i = 0; i < input_port_count; i++) {
 
 		Vector<String> arr = input_strings[i].split(",");
+		ERR_FAIL_INDEX(0, arr.size());
```
I think, there should be check for exactly 3 arguments instead (ERR_FAIL_COND(arr.size() != 3))

Dentrax replied:
Actually, I did that on purpose. If I checked in _!= 3_ way, we cannot understand exactly which index of the error is in the editor. But only because I check the index, I can see that the error is exactly in _nth_ index.

Chaosus replied:
Ah, nvm then

akien-mga replied:
`ERR_FAIL_INDEX` is meant to compare a user-passed index with an array size, not the other way around.
This should simply be `ERR_FAIL_COND(arr.size() != 3);` as suggested.

Users won't get a lot more information from `Index 2=2 out of size (arr.size()=3)` than a plain `Condition ' arr.size() != 3 ' is true.` IMO. The latter is more explicit about the fact that something should have a size of 3.

Chaosus commented on scene/resources/visual_shader.cpp:
```diff
@@ -1792,8 +1795,10 @@ void VisualShaderNodeGroupBase::set_outputs(const String &p_outputs) {
 	for (int i = 0; i < output_port_count; i++) {
-
 		Vector<String> arr = output_strings[i].split(",");
+		ERR_FAIL_INDEX(0, arr.size());
```
Same as above

akien-mga commented on drivers/gles3/rasterizer_storage_gles3.cpp:
```diff
@@ -4635,7 +4635,6 @@ Transform2D RasterizerStorageGLES3::multimesh_instance_get_transform_2d(RID p_mu
 }
 
 Color RasterizerStorageGLES3::multimesh_instance_get_color(RID p_multimesh, int p_index) const {
-
```
It would be better not to remove this line, as it marks this file as modified by this commit even though there is no functional changes.

Dentrax replied:
Oh, I did not change that file, it happened accidentally somehow. :)

akien-mga replied:
If you can undo the change next time you amend the commit, that would be great :)
But it's not a big deal either if it stays this way.

akien-mga commented on modules/bullet/rigid_body_bullet.cpp:
```diff
@@ -866,7 +866,7 @@ void RigidBodyBullet::on_enter_area(AreaBullet *p_area) {
 	if (p_area->is_spOv_gravityPoint()) {
 		++countGravityPointSpaces;
-		assert(0 < countGravityPointSpaces);
+		ERR_FAIL_COND(countGravityPointSpaces < 0);
```
If we want to match the previous code it should be `<= 0`. Or is allowing the `0` cases part of the fix?

Dentrax replied:
I think `< 0` is fine. Because of, that initialized with value of `0` at line `274`. Also it checked with `< 0` at line `379` as well. Why do you think that should be `<= 0`?

akien-mga replied:
Well I ask because the previous code was asserting on `countGravityPointSpaces > 0`, i.e failing on `countGravityPointSpaces <= 0`. I'm not saying the previous code was correct, just that this changes the logic so I wanted to make sure it's done on purpose.

Dentrax replied:
Oh, you are right. I missed that point. :)

akien-mga replied:
So, this is still changing the logic.
`assert` checks that a condition is true, and `ERR_FAIL_COND` checks that a condition is false, so if you want to change `assert` to `ERR_FAIL_COND`, you should negate the condition. The negation of `0 < countGravityPointSpaces` (i.e. `countGravityPointSpaces > 0` without the Java-ism) is `countGravityPointSpaces <= 0`.

Dentrax replied:
I'm always confused when I'm transforming between `assert` and `cond`. Thank you! Fixed.

akien-mga commented on modules/gdscript/gdscript_function.cpp:
```diff
@@ -1549,7 +1549,9 @@ Variant GDScriptFunction::call(GDScriptInstance *p_instance, const Variant **p_a
 		String err_file;
 		if (p_instance)
-			err_file = p_instance->script->path;
+			if(p_instance->script->path != "") {
```
Style issue, there should be a space after `if`.

Dentrax replied:
Fixed, and I am curious about something. Why we not use a static reference to check a string is empty or null instead of creating new string everytime like `""`?

akien-mga replied:
You can use `!path.empty()` if you want. It would be interesting to benchmark what difference it makes, if any.

akien-mga commented on scene/3d/visual_instance.cpp:
```diff
@@ -60,7 +60,7 @@ void VisualInstance::_notification(int p_what) {
 				VisualServer::get_singleton()->instance_attach_skeleton( instance, skeleton->get_skeleton() );
 			*/
-
+			ERR_FAIL_COND(get_world().is_null());
```
The fix seems fine, but in what case do we receive `NOTIFICATION_ENTER_WORLD` without a valid `get_world()`?

Dentrax replied:
I don't know exactly but somehow the function of `Ref<World> Viewport::find_world() const` returns _else_ situation. Maybe it returns `Ref<World>()` if `is_inside_world()` is _false_ or `data.viewport` is `null`. How can we test that `get_world()` why returns _null_ or _empty_ with a better way? What is your opinion?

akien-mga commented on scene/main/http_request.cpp:
```diff
@@ -159,16 +159,20 @@ void HTTPRequest::cancel_request() {
-		memdelete(thread);
-		thread = NULL;
+
+		if(thread) {
```
Style issue.

akien-mga commented on scene/main/node.cpp:
```diff
@@ -71,6 +71,8 @@ void Node::_notification(int p_notification) {
 		} break;
 		case NOTIFICATION_ENTER_TREE: {
+			ERR_FAIL_COND(!get_viewport());
+			ERR_FAIL_COND(!get_tree());
```
It also seems weird that any of those two can happen, `NOTIFICATION_ENTER_TREE` is emitted when the node enters the tree, which means that there should be a tree (and thus a root, so a viewport).

Dentrax replied:
I am suprised too. But if I don't check this way, the editor crashes. I think thats very rare condition.

akien-mga commented on servers/audio/effects/audio_effect_record.cpp:
```diff
@@ -216,6 +216,9 @@ Ref<AudioStreamSample> AudioEffectRecord::get_recording() const {
 	PoolVector<uint8_t> dst_data;
 
+	ERR_FAIL_COND_V(current_instance.is_null(), NULL)
+	ERR_FAIL_COND_V(current_instance->recording_data.size(), NULL)
```
Missing semicolons.

akien-mga commented:
For the VisualShader bugs, I prefer the approach in #29987 as it does most of the input validation before running into issues in the for loops. So once that one is merged, you'll have to amend/rebase your commit to remove the conflicting changes.

akien-mga commented:
It would be nice to amend the commit log too to be slightly more explicit, even though I agree that it's hard to give a good description of such a commit. Something like "Add checks to prevent out of bounds indexing".

Dentrax commented:
@akien-mga 

Thanks for the suggestion. I will take heed of to commit messages for my next work. :)

To merge to @Chaosus 's PR, I did this way: 

`git checkout upstream/master -- .\scene\resources\visual_shader.cpp` and then `git add -u` then `git commit --amend` Is that correct? There are some things about git that I don't know yet. Sorry for that. I don't want that code to be like I changed it. But now, `visual_shader.cpp` seems that I have changed. How can i fix that?

Faless commented on scene/main/http_request.cpp:
```diff
@@ -159,16 +159,21 @@ void HTTPRequest::cancel_request() {
-		memdelete(thread);
-		thread = NULL;
+
+		if (thread) {
```
I think a better solution would be to fix the condition in `set_use_threads` that can cause the value to change before the request has finished, thus causing `use_threads` to be `true` but `thread` to be null. (`_request_done`/`cancel_request` is called deferred when status become disconnected during poll, thus causing a frame to be in the situation where `status == STATUS_DISCONNECTED` but `requesting == true`)

```diff
diff --git a/scene/main/http_request.cpp b/scene/main/http_request.cpp
index e65314644..d6392d62d 100644
--- a/scene/main/http_request.cpp
+++ b/scene/main/http_request.cpp
@@ -425,7 +425,7 @@ void HTTPRequest::_notification(int p_what) {
 
 void HTTPRequest::set_use_threads(bool p_use) {
 
-       ERR_FAIL_COND(status != HTTPClient::STATUS_DISCONNECTED);
+       ERR_FAIL_COND(requesting);
        use_threads = p_use;
 }
```

Dentrax replied:
I think, thats a bad idea to remove that `STATUS_DISCONNECTED` because someone can be easily confused at that line. Intead, should we check the `requesting` status that way? @akien-mga 

```
if (!requesting) {
	use_threads = p_use;
}
```

Faless replied:
> Isn't this exactly the change I proposed above in the diff?

Oh, I see the difference, keeping both COND, sorry. Yeah, we might keep both conditions. I think it needs to print an error though if you try to change it while requesting.

Faless replied:
I think I found the underlying issue. See #30218

akien-mga replied:
The changes in this file can now be removed since #30218 is merged.

Dentrax replied:
Looks like @Faless 's way is more effective solution. I removed that changes.

akien-mga commented on modules/bullet/rigid_body_bullet.cpp:
```diff
@@ -888,7 +888,7 @@ void RigidBodyBullet::on_exit_area(AreaBullet *p_area) {
 		if (p_area->is_spOv_gravityPoint()) {
 			--countGravityPointSpaces;
-			assert(0 <= countGravityPointSpaces);
+			ERR_FAIL_COND(countGravityPointSpaces <= 0);
```
This changes logic too. The asserts here and at line 869 were different, and assuming they were correct, the new `ERR_FAIL_COND` should respect that logic. `countGravityPointSapces` is initialized to 0, so I guess 0 should be a valid value.

Dentrax replied:
Oh, you are right. I missed it.

akien-mga replied:
Same as above, but this time it's the other way around.

The original condition was `countGravityPointSpaces >= 0`, so when negated, it should be `countGravityPointSpaces < 0`.

Dentrax replied:
Fixed

akien-mga commented on modules/gdscript/gdscript_function.cpp:
```diff
@@ -1549,7 +1549,9 @@ Variant GDScriptFunction::call(GDScriptInstance *p_instance, const Variant **p_a
 		String err_file;
 		if (p_instance)
-			err_file = p_instance->script->path;
+			if (p_instance->script->path != "") {
```
What's the point of this check actually? If `p_instance->script->path == ""`, then `err_file` will get an empty String assigned, which is fine as it's its current value anyway.

Didn't you mean to check `p_instance->script->is_valid()` to prevent a null pointer dereference?

Dentrax replied:
Interesting part is, the bug was fixed that way. Now I am adding the `is_valid()` checks.

akien-mga commented on scene/2d/canvas_item.cpp:
```diff
@@ -1075,6 +1075,7 @@ Vector2 CanvasItem::make_canvas_position_local(const Vector2 &screen_point) cons
 Ref<InputEvent> CanvasItem::make_input_local(const Ref<InputEvent> &p_event) const {
 
 	ERR_FAIL_COND_V(!is_inside_tree(), p_event);
+	ERR_FAIL_COND_V(p_event.is_null(), NULL);
```
It's mostly cosmetic, but I'd say you could return `p_event` here to be consistent with the previous line and overall logic: return the input on failure. Even though here `p_event` *is* NULL so it's the same in the end.

akien-mga commented:
> There are some things about git that I don't know yet. Sorry for that. I don't want that code to be like I changed it. But now, visual_shader.cpp seems that I have changed. How can i fix that?

You can probably remove those changes with `git rebase upstream/master`.

Dentrax commented:
@akien-mga Could you review again, please? :)

akien-mga commented:
Thanks!

akien-mga commented on scene/main/node.cpp:
```diff
@@ -840,6 +844,8 @@ bool Node::is_processing_internal() const {
 void Node::set_process_priority(int p_priority) {
 	data.process_priority = p_priority;
 
+	ERR_FAIL_COND(!data.tree);
```
This now causes an error when loading a project in the editor:
```
ERROR: set_process_priority: Condition ' !data.tree ' is true.
   At: scene/main/node.cpp:847.
```

Dentrax replied:
Hmm, what condition are you testing to get this error? If you hadn't taken this error, editor would crash in one of the 4 if structures. According to your [comment](https://github.com/godotengine/godot/pull/29980#discussion_r296474512), is there a condition that must not enter the any of structures?

akien-mga replied:
Just open the editor on any project:
```
$ mkdir ~/tmp/new_project
$ cd ~/tmp/new_project
$ touch project.godot
$ godot -e
Godot Engine v3.2.dev.custom_build.266dc732b - https://godotengine.org
Found discrete GPU, setting DRI_PRIME=1 to use it.
Note: Set DRI_PRIME=0 in the environment to disable Godot from using the discrete GPU.
OpenGL ES 3.0 Renderer: AMD VEGAM (DRM 3.30.0, 5.1.14-desktop-1.mga7, LLVM 8.0.0)
 
ERROR: set_process_priority: Condition ' !data.tree ' is true.
   At: scene/main/node.cpp:847.
```

Dentrax replied:
I tried your way and also i created new project from editor. And I didn't get that error on version `v3.2.dev.custom_build. 7d8d337b2`.

akien-mga replied:
@bojidar-bg fixed it in #30254, so no worry :)
